File: airesources/Ruby/hlt/player.rb
require 'ship'

class Player
  attr_reader :id

  def initialize(player_id, ships = {})
    @ships = ships
    @id = player_id
  end

  def ships
    @ships.values
  end

  def ship(ship_id)
    @ships[ship_id]
  end

  # Parse an entire user input string from the Halite engine for all users.
  # tokens: The input string as a list of str from the Halite engine.
  # return: The parsed players in the form of player hash, and remaining tokens
  def self.parse(tokens)
    count_of_players = Integer(tokens.shift)
    players = {}

    count_of_players.times do
      player_id, player, tokens = parse_single(tokens)
      players[player_id] = player
    end

    return players, tokens
  end

  # Parse one user given an input string from the Halite engine.
  # tokens: The input string as an array of strins.
  # return: The parsed player id, player object, and remaining tokens
  def self.parse_single(tokens)
    player_id = Integer(tokens.shift)
    ships, tokens = Ship.parse(player_id, tokens)
    player = Player.new(player_id, ships)
    return player_id, player, tokens
  end

  def to_s
    "Player #{id} with ships #{ships}"
  end
end

File: airesources/Julia/hlt/player.jl
struct Player
    id::String
    ships::Dict{String, Ship}
end
Player(id::String) = Player(id, Dict{String, Ship}())

"""
    get_ship(player, ship_id::String)

Returns the ship designated by `ship_id` belonging to this user.
"""
get_ship(player::Player, ship_id::String) = player.ships[ship_id]

"""
    all_ships(player)

Returns iterator over all ships which belong to the user.
"""
all_ships(player::Player) = values(player.ships)
